### PR TITLE
Total Mechanical Energy

### LINKED ISSUE #16201: Function to return Total Energy in physics/mechanics
At present, there is are functions to calculate kinetic energy and potential energy of a rigid body in ```sympy/physics/mechanics/rigidbody.py``` but no function to return the total energy of the rigid body. I propose the following PR. Please let me know your views about it.

### PR DESCRIPTION
**Pull request by sidhu1012**
#### References to other Issues or PRs

fixes #16201 
Using PR: #18838
#### Brief description of what is fixed or changed
Added new function to particle.py

#### Other comments
TE = P.E + K.E

#### Release Notes

* physics.mechanics
  * Added new function to calculate total energy of a particle.

### CODE REVIEW

**Comment by sidhu1012** [comment]
What went wrong?🤔

**Comment by czgdp1807** [comment]
See, https://travis-ci.org/github/sympy/sympy/jobs/722265397

File: sympy/physics/mechanics/tests/test_particle.py [hunk 1]
```diff
@@ -4,6 +4,8 @@
 from sympy.testing.pytest import raises
 
 
+
+
```
Comment (czgdp1807) [inline sympy/physics/mechanics/tests/test_particle.py:8]:
```suggestion
```

File: sympy/physics/mechanics/tests/test_particle.py [hunk 2]
```diff
@@ -54,3 +57,4 @@ def test_parallel_axis():
     Ip_expected = inertia(N, m * b**2, m * a**2, m * (a**2 + b**2),
                           ixy=-m * a * b)
     assert Ip == Ip_expected
+
```
Comment (czgdp1807) [inline sympy/physics/mechanics/tests/test_particle.py:60]:
```suggestion
```

File: sympy/physics/mechanics/particle.py [hunk 3]
```diff
@@ -240,6 +240,37 @@ def set_potential_energy(self, scalar):
...
+
+        frame : ReferenceFrame
+            The Particle's velocity is typically defined with respect to
+            an inertial frame but any relevant frame in which the velocity is
+            known can be supplied.
+
+         Examples
+         ========
+
+        >>> from sympy.physics.mechanics import Particle, Point, ReferenceFrame
```
Comment (eric-wieser) [inline sympy/physics/mechanics/particle.py:260]: Indent is broken

```suggestion
        Examples
        ========

        >>> from sympy.physics.mechanics import Particle, Point, ReferenceFrame
```

File: sympy/physics/mechanics/particle.py [hunk 4]
```diff
@@ -240,6 +240,37 @@ def set_potential_energy(self, scalar):
...
+        >>> from sympy import symbols
+        >>> m, v, g, h = symbols('m v g h')
+        >>> N = ReferenceFrame('N')
+        >>> O = Point('O')
+        >>> P = Particle('P', O, m)
+        >>> P.potential_energy = m * g * h
+        >>> P.point.set_vel(N, v * N.y)
+        >>> E = total_mechanical_energy(N)
+
```
Comment (eric-wieser) [inline sympy/physics/mechanics/particle.py:269]:
```suggestion
        >>> total_mechanical_energy(N)
        <show what the output is>
```

File: sympy/physics/mechanics/tests/test_particle.py [hunk 5]
```diff
@@ -42,6 +42,7 @@ def test_particle():
     assert p.kinetic_energy(
         N) in [m2*(v1**2 + v2**2 + v3**2)/2,
         m2 * v1**2 / 2 + m2 * v2**2 / 2 + m2 * v3**2 / 2]
+    assert(p.total_mechanical_energy(N)) == g*h*m + m2*v1**2/2 + m2*v2**2/2 + m2*v3**2/2
```
Comment (eric-wieser) [inline sympy/physics/mechanics/tests/test_particle.py:45]:
```suggestion
    assert p.total_mechanical_energy(N) == g*h*m + m2*v1**2/2 + m2*v2**2/2 + m2*v3**2/2
```

**Comment by sidhu1012** [comment]
Now what's wrong?

image


Why this test is failing

**Comment by sidhu1012** [comment]
What all test does travis ci do?

**Comment by sidhu1012** [comment]
Finally passed all tests!

**Comment by sidhu1012** [comment]
Is total energy of rigid body  given by total kinetic energy + potential energy?

**Comment by sidhu1012** [comment]
@moorepants @eric-wieser @czgdp1807 Can you pls have a look to this PR and review it. If it looks good I would like to add total energy function for rigid body too.

**Comment by moorepants** [comment]
It seems to me that you are copying from this PR: https://github.com/sympy/sympy/pull/18838/

If you want to build on someone else's work, that is fine. But please retain their commits so they get credit.

People have suggested this function before and I personally find it a bit unnecessary.

Why do we need a method:

```
particle.total_energy(frame)
```

when we can just type:

```
particle.kinetic_energy(frame) + particle.potential_energy()
```

It saves some key strokes, I guess. It doesn't hurt to add it, so I won't block, but please address the copying.

**Comment by sidhu1012** [comment]
> It seems to me that you are copying from this PR: #18838
> 
> If you want to build on someone else's work, that is fine. But please retain their commits so they get credit.
> 
> People have suggested this function before and I personally find it a bit unnecessary.
> 
> Why do we need a method:
> 
> ```
> particle.total_energy(frame)
> ```
> 
> when we can just type:
> 
> ```
> particle.kinetic_energy(frame) + particle.potential_energy()
> ```
> 
> It saves some key strokes, I guess. It doesn't hurt to add it, so I won't block, but please address the copying.

I didn't see any other PR but if it's same i'll surely mention that.

**Comment by moorepants** [comment]
> I didn't see any other PR but if it's same i'll surely mention that.

Some of the text in the docstring is exactly the same as the other PR. Maybe there is another explanation, but that seems copied to me.

**Comment by sidhu1012** [comment]
> > I didn't see any other PR but if it's same i'll surely mention that.
> 
> Some of the text in the docstring is exactly the same as the other PR. Maybe there is another explanation, but that seems copied to me.

It wasn't copied , wrote it myself without any idea of other similar PR.

But I think total energy function should - be there because somw people may not know that Total energy = Kinetic energy + Potential
They just need toh know total energy

In such cases isn't it good to have totoal energy function?

**Comment by oscarbenjamin** [comment]
> without any idea of other similar PR.

I find this implausible after looking at the diff for both PRs.

In any case there was prior work on this so I think that the best thing would be for you to look at #16201 and help to finish that. You can base a new PR on top of the commits already there.

> somw people may not know that Total energy =

I find it hard to imagine that someone with that level of mechanics would be able to make use of sympy for this sort of calculation.

**Comment by sidhu1012** [comment]
> > without any idea of other similar PR.
> 
> I find this implausible after looking at the diff for both PRs.
> 
> In any case there was prior work on this so I think that the best thing would be for you to look at #16201 and help to finish that. You can base a new PR on top of the commits already there.
> 
> > somw people may not know that Total energy =
> 
> I find it hard to imagine that someone with that level of mechanics would be able to make use of sympy for this sort of calculation.

Okay😊

**Comment by eric-wieser** [comment]
> I find this implausible after looking at the diff for both PRs.

The docstring and example is almost entirely copied from `kinetic_energy` in both PRs - so I think it's quite plausible that they were developed independently.

**Comment by oscarbenjamin** [comment]
> I think it's quite plausible

Okay, good point.

In any case I still think that it would be best to begin by reviewing or attempting to finish the other PR rather than opening a new one.

**Comment by oscarbenjamin** [comment]
@sidhu1012 I think it would be better to make a PR starting from #18838 (keeping the commits intact) and making changes from there as it has more than this one and was already reviewed quite a bit. I would start by asking on that PR whether all of the changes are wanted though.

**Comment by sidhu1012** [comment]
@oscarbenjamin  Yes but  that PR is idle for more than a year

**Comment by oscarbenjamin** [comment]
That's why it would be good for someone to finish it off rather than start a new PR to go over the same work.

**Comment by sidhu1012** [comment]
> That's why it would be good for someone to finish it off rather than start a new PR to go over the same work.

Okay if you say so.

**Comment by oscarbenjamin** [comment]
The issue is now closed: https://github.com/sympy/sympy/issues/16201#issuecomment-702692683

@sidhu1012 thank you for working on this but I don't think that we need a function for calculating the total energy.

**Pull request closed without merging**